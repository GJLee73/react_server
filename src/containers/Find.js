// 아이디 비밀번호 찾기

import React from 'react';
import {FindId,FindPwd} from '../components/auth'
import {connect} from 'react-redux';
import {Link} from 'react-router-dom'
import {loginRequest,findId,findPwd} from 'modules/authentication';

class Find extends React.Component {
    constructor(props) {
        super(props);
        this.state = { select:true, gottenId:''} //true ='id' false='pwd'
        this.handleSelect=this.handleSelect.bind(this)
        this.sendEmail=this.sendEmail.bind(this)
    }
    handleSelect(select){
        if(select){
            this.setState({...this.state, select:true})
        }else{
            this.setState({...this.state, select:false})
        }
    }
    /** 서버에 key와 이메일을 보내 아이디와 비밀번호를 찾을 수 있게한다.
     * @param {string} key - id 찾기는 name값 비밀번호 찾기는 Id가 담겨있다.
     * @param {string} email -email
     */
    sendEmail(key,email){
        if(this.state.select){//id찾기면
            console.log('1',key,email)
            this.props.findId(key,email).then(()=>{
                this.setState({gottenId:this.props.gottenId})
            }
                
            )
            .catch(()=>{alert('이름과 이메일이 틀립니다.')})
            
        }else{//비밀번호찾기면
            this.props.findPwd(key,email).then(()=>{
               alert('이메일에 비밀번호 변경 링크가 전송되었습니다.')
               this.props.history.push('/login')
            })
            .catch(()=>{alert('아이디와 이메일이 틀립니다.')})
        }

    }
    render() {
        return (
            <div>
                <div className = 'container'>
            <div className = 'card'>
            <div className="header blue white-text center">
                            <div className="card-content">
                            <i onClick={()=>{this.handleSelect(true)}} style={{cursor:'pointer'}}>아이디찾기</i>
                            /
                            <i onClick={()=>{this.handleSelect(false)}} style={{cursor:'pointer'}}>비밀번호찾기</i>
                            </div>
            </div>
            {this.state.gottenId===''? (this.state.select? <FindId    sendEmail={this.sendEmail}/>:
            <FindPwd  sendEmail={this.sendEmail}/> ): <div>
                {this.state.gottenId}
                <Link to ='/login'>로그인하러 가기</Link>
            </div>}
            
            
           
            </div> 
            </div>
        </div>
      
        )
    }
}
const mapStateToProps = (state) => {
	return {
        status: state.authentication.getIn(['login','status']),
        gottenId: state.authentication.getIn(['findId','gottenId'])
	};
};

const mapDispatchToProps = (dispatch) => {
	return {
		loginRequest: (id, pw)=>{
            return dispatch(loginRequest(id,pw))},
        findId:(name,email)=>{
            return dispatch(findId(name,email))},
        findPwd:(id,email)=> dispatch(findPwd(id,email))
        
		}
};

export default connect(mapStateToProps, mapDispatchToProps)(Find);