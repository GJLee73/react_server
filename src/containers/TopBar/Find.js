/**
 * 아이디 비밀번호 찾기
 * @author G1
 * @logs // 18.2.26
 */

import React from 'react';
import {FindId,FindPwd} from '../../components/auth'
import { Modal, ModalHeader, ModalBody, ModalFooter } from 'reactstrap';


export class Find extends React.Component {
    constructor(props) {
        super(props);
        this.state = { select:true, gottenId:''} //true ='id' false='pwd'
        this.handleSelect=this.handleSelect.bind(this)
        this.sendEmail=this.sendEmail.bind(this)
    }
    /**
     * 아이디 찾기인지, 비밀번호 찾기인지 선택할 수 있게 한다.
     * @param {boolean} select 
     */
    handleSelect(select){
        if(select){
            this.setState({...this.state, select:true})
        }else{
            this.setState({...this.state, select:false})
        }
    }
    /** 서버에 key와 이메일을 보내 아이디와 비밀번호를 찾을 수 있게한다.
     * @param {string} key - id 찾기는 name값 비밀번호 찾기는 Id가 담겨있다.
     * @param {string} email -email
     */
    sendEmail(key,email){
        if(this.state.select){//id찾기면
            console.log('1',key,email)
          return this.props.findId(key,email).then(()=>{
                this.setState({gottenId:this.props.gottenId})
            })
            .catch(()=>{
                console.log(1)
                alert('이름과 이메일이 틀립니다.')})
            
        }else{//비밀번호찾기면
           return this.props.findPwd(key,email).then(()=>{
               alert(this.props.message)
               this.props.history.push('/login')
            })
            .catch(()=>{alert(this.props.message)})
        }

    }
    render() {
        return (
            <Modal isOpen={this.props.modal} toggle={this.props.toggle} className={this.props.className}style={{marginTop:'100px',width:'350px'}}>
			<div style={{paddingTop:"40px",paddingLeft:"50px"}} className="d-flex flex-row">

                            <i onClick={()=>{this.handleSelect(true)}} style={{cursor:'pointer'}}>아이디찾기</i>
                            /
                            <i onClick={()=>{this.handleSelect(false)}} style={{cursor:'pointer'}}>비밀번호찾기</i>
			</div>
			<ModalBody style={{paddingTop:"0",paddingLeft:"30px"}}>
            {this.state.gottenId===''? (this.state.select? <FindId    sendEmail={this.sendEmail}/>:
            <FindPwd  sendEmail={this.sendEmail}/> ): <div>
                {this.state.gottenId}
            </div>}
			</ModalBody>
			<ModalFooter>
			<div className="card-content">
				<div className="right d-flex " >
                	<div className="text-secondary" onClick={()=>{this.props.changeType('login')}}>로그인하기</div>	
				</div>
			</div>
			</ModalFooter>
			
			</Modal>
        )
    }
}

export default Find