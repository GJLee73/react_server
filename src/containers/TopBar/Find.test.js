import React from 'react';
import {shallow} from 'enzyme'
import {Find} from './Find'

describe('Find test',()=>{
    var wrapper = shallow(<Find/>)
    var history ={push:jest.fn()}
    global.alert=jest.fn()
    describe('function test',()=>{
        describe('handleSelect test',()=>{
            it('select==true',()=>{
                wrapper.instance().handleSelect(true)
                expect(wrapper.state().select).toBe(true)
            })
            it('select==false',()=>{
                wrapper.instance().handleSelect(false)
                expect(wrapper.state().select).toBe(false)
            })
        })
        describe('sendEmail test',()=>{
            it('id 찾기이면 select==true',async()=>{
                wrapper.setProps({findId:jest.fn(()=>Promise.resolve()),gottenId:'g1'})
                wrapper.setState({select:true})
                await wrapper.instance().sendEmail()
                expect(wrapper.state('gottenId')).toBe('g1')
            })
            it('id 찾기를 실패한다면',async()=>{
                wrapper.setProps({findId:jest.fn(()=>Promise.reject())})
                await wrapper.instance().sendEmail()
                expect(global.alert.mock.calls[0][0]).toBe('이름과 이메일이 틀립니다.')
            })
            it('비밀번호 찾기이면 select==false',async()=>{
                wrapper.setState({select:false})
                wrapper.setProps({findPwd:()=>Promise.resolve(),history,message:'이메일에 비밀번호 변경 링크가 전송되었습니다.'})
                await wrapper.instance().sendEmail()
                expect(global.alert.mock.calls[1][0]).toBe('이메일에 비밀번호 변경 링크가 전송되었습니다.')
                expect(history.push).toHaveBeenCalledTimes(1)
            })
            it('findPwd가 실패하면 ',async()=>{
                wrapper.setProps({findPwd:()=>Promise.reject(),message:'아이디와 이메일이 틀립니다.'})
                await wrapper.instance().sendEmail()
                expect(global.alert.mock.calls[2][0]).toBe('아이디와 이메일이 틀립니다.')
                expect(history.push).toHaveBeenCalledTimes(1)
            })
        })
    })
    describe('snapshot test',()=>{
        it('id찾기일때',()=>{
            wrapper.setState({select:true,gottenId:''})
            expect(wrapper).toMatchSnapshot()
        })
        it('비밀번호 찾기일때',()=>{
            wrapper.setState({select:false})
            expect(wrapper).toMatchSnapshot()
        })
        it('아이디를 찾아서 아이디가 나타나면',()=>{
            wrapper.setState({gottenId:'g1'})
            expect(wrapper).toMatchSnapshot()
        })
    })
})