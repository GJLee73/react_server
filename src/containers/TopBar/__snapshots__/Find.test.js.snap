// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Find test snapshot test id찾기일때 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropTransition={
    Object {
      "mountOnEnter": true,
      "timeout": 150,
    }
  }
  centered={false}
  fade={true}
  isOpen={false}
  keyboard={true}
  modalTransition={
    Object {
      "timeout": 300,
    }
  }
  onClosed={[Function]}
  onOpened={[Function]}
  role="dialog"
  style={
    Object {
      "marginTop": "100px",
      "width": "350px",
    }
  }
  zIndex={1050}
>
  <div
    className="d-flex flex-row"
    style={
      Object {
        "paddingLeft": "50px",
        "paddingTop": "40px",
      }
    }
  >
    <i
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      아이디찾기
    </i>
    /
    <i
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      비밀번호찾기
    </i>
  </div>
  <ModalBody
    style={
      Object {
        "paddingLeft": "30px",
        "paddingTop": "0",
      }
    }
    tag="div"
  >
    <FindId
      sendEmail={[Function]}
    />
  </ModalBody>
  <ModalFooter
    tag="div"
  >
    <div
      className="card-content"
    >
      <div
        className="right d-flex "
      >
        <div
          className="text-secondary"
          onClick={[Function]}
        >
          로그인하기
        </div>
      </div>
    </div>
  </ModalFooter>
</Modal>
`;

exports[`Find test snapshot test 비밀번호 찾기일때 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropTransition={
    Object {
      "mountOnEnter": true,
      "timeout": 150,
    }
  }
  centered={false}
  fade={true}
  isOpen={false}
  keyboard={true}
  modalTransition={
    Object {
      "timeout": 300,
    }
  }
  onClosed={[Function]}
  onOpened={[Function]}
  role="dialog"
  style={
    Object {
      "marginTop": "100px",
      "width": "350px",
    }
  }
  zIndex={1050}
>
  <div
    className="d-flex flex-row"
    style={
      Object {
        "paddingLeft": "50px",
        "paddingTop": "40px",
      }
    }
  >
    <i
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      아이디찾기
    </i>
    /
    <i
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      비밀번호찾기
    </i>
  </div>
  <ModalBody
    style={
      Object {
        "paddingLeft": "30px",
        "paddingTop": "0",
      }
    }
    tag="div"
  >
    <FindPwd
      sendEmail={[Function]}
    />
  </ModalBody>
  <ModalFooter
    tag="div"
  >
    <div
      className="card-content"
    >
      <div
        className="right d-flex "
      >
        <div
          className="text-secondary"
          onClick={[Function]}
        >
          로그인하기
        </div>
      </div>
    </div>
  </ModalFooter>
</Modal>
`;

exports[`Find test snapshot test 아이디를 찾아서 아이디가 나타나면 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropTransition={
    Object {
      "mountOnEnter": true,
      "timeout": 150,
    }
  }
  centered={false}
  fade={true}
  isOpen={false}
  keyboard={true}
  modalTransition={
    Object {
      "timeout": 300,
    }
  }
  onClosed={[Function]}
  onOpened={[Function]}
  role="dialog"
  style={
    Object {
      "marginTop": "100px",
      "width": "350px",
    }
  }
  zIndex={1050}
>
  <div
    className="d-flex flex-row"
    style={
      Object {
        "paddingLeft": "50px",
        "paddingTop": "40px",
      }
    }
  >
    <i
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      아이디찾기
    </i>
    /
    <i
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      비밀번호찾기
    </i>
  </div>
  <ModalBody
    style={
      Object {
        "paddingLeft": "30px",
        "paddingTop": "0",
      }
    }
    tag="div"
  >
    <div>
      g1
    </div>
  </ModalBody>
  <ModalFooter
    tag="div"
  >
    <div
      className="card-content"
    >
      <div
        className="right d-flex "
      >
        <div
          className="text-secondary"
          onClick={[Function]}
        >
          로그인하기
        </div>
      </div>
    </div>
  </ModalFooter>
</Modal>
`;
